cas-compute/src/symbolic/derivative.rs

```diff
@@ -0,0 +1,268 @@
+use rug::Integer;
+
+use crate::primitive::{float, int};
+
+use super::expr::Primary;
+use super::Expr;
+
+// Something that is not trivially zero - intended to clean up ASTs and not mathetmatically
+// rigorous
+fn non_zero(e: &Expr) -> bool {
+    match e {
+        Expr::Primary(Primary::Float(f)) => {
+            !f.is_zero()
+        },
+        Expr::Primary(Primary::Integer(i)) => {
+            !i.is_zero()
+        },
+        Expr::Add(sum) => {
+            sum.iter().any(non_zero)
+        },
+        Expr::Mul(mul) => {
+            mul.iter().all(non_zero)
+        }
+        _ => true
+    }
+}
+
+// Something that is not trivially one - intended to clean up ASTs and not mathetmatically
+// rigorous
+fn non_unity(e: &Expr) -> bool {
+    match e {
+        Expr::Primary(Primary::Float(f)) => !f.eq(&float(1)),
+        Expr::Primary(Primary::Integer(i)) => !i.eq(&int(1)),
+        Expr::Mul(exprs) => {
+            exprs.iter().any(non_unity)
+        },
+        Expr::Exp(expr, expr1) => {
+            non_zero(expr1) && non_unity(expr)
+        },
+        _ => true
+    }
+}
+
+#[derive(Default)]
+struct MultBuilder(Vec<Expr>);
+
+impl From<MultBuilder> for Expr {
+    fn from(value: MultBuilder) -> Self {
+        if value.0.len() > 1 {
+            Expr::Mul(value.0)
+        } else if value.0.len() == 1 {
+            value.0[0].clone()
+        } else {
+            Expr::Primary(Primary::Integer(int(1)))
+        }
+    }
+}
+
+impl MultBuilder {
+    fn mult(&mut self, e: Expr) {
+        if !non_zero(&e) || self.0.get(0).is_some_and(|e| !non_zero(e)) {
+            self.0 = vec![Expr::Primary(Primary::Integer(int(0)))];
+            return;
+        }
+
+        if non_unity(&e) {
+            self.0.push(e)
+        }
+    }
+}
+
+#[derive(Default)]
+struct SumBuilder(Vec<Expr>);
+
+impl From<SumBuilder> for Expr {
+    fn from(value: SumBuilder) -> Self {
+        if value.0.len() > 1 {
+            Expr::Add(value.0)
+        } else if value.0.len() == 1 {
+            value.0[0].clone()
+        } else {
+            Expr::Primary(Primary::Integer(int(0)))
+        }
+    }
+}
+
+impl SumBuilder {
+    fn add(&mut self, e: Expr) {
+        if non_zero(&e) {
+            self.0.push(e)
+        }
+    }
+}
+
+fn sum_rule(exprs: &[Expr], var: &str) -> Result<Expr, SymbolicDerivativeError> {
```

Comment:
```suggestion
/// `(f + g)' = f' + g'`
fn sum_rule(exprs: &[Expr], var: &str) -> Result<Expr, SymbolicDerivativeError> {
```

Documentation for these rules can be a lot simpler.